File: isabella.ml/src/canon/confidential_merkle.ml
(** Cryptographic Merkle helpers for confidential note commitments.

    This module is the OCaml runtime companion to Canon/ZK/Authenticated_Merkle.
    It intentionally stays separate from Confidential_transaction while that
    transaction verifier still uses the algebraic ledger scaffold. *)

type digest = string

type membership_proof = {
  merkle_index : int;
  merkle_root : digest;
  merkle_siblings : digest list;
  merkle_directions : bool list;
}

let merkle_dst = "ISABELLA-CT-MERKLE-v1"

let merkle_leaf_tag = 0
let merkle_node_tag = 1
let merkle_empty_tag = 2

let digest_hex bytes =
  String.concat "" (List.map (Printf.sprintf "%02x") bytes)

let digest preimage =
  digest_hex (Repeated_fs.sha3_256 preimage)

let encode_i64 value =
  Repeated_fs.int64_le_bytes (Int64.of_int value)

let encode_int_vec values =
  encode_i64 (List.length values) @ List.concat (List.map encode_i64 values)

let preimage tag body =
  Repeated_fs.string_bytes merkle_dst @ encode_i64 tag @ body

let encode_leaf commitment =
  digest_hex (preimage merkle_leaf_tag (encode_int_vec commitment))

let encode_empty width =
  if width < 0 then invalid_arg "empty width must be non-negative";
  digest_hex (preimage merkle_empty_tag (encode_i64 width))

let hex_value c =
  match c with
  | '0' .. '9' -> Some (Char.code c - Char.code '0')
  | 'a' .. 'f' -> Some (10 + Char.code c - Char.code 'a')
  | _ -> None

let hex_to_bytes hex =
  if String.length hex <> 64 then None
  else
    let rec loop i acc =
      if i = String.length hex then Some (List.rev acc)
      else
        match hex_value hex.[i], hex_value hex.[i + 1] with
        | Some hi, Some lo -> loop (i + 2) (((hi lsl 4) lor lo) :: acc)
        | _ -> None
    in
    loop 0 []

let encode_node left right =
  match hex_to_bytes left, hex_to_bytes right with
  | Some left_bytes, Some right_bytes ->
      digest_hex
        (preimage
           merkle_node_tag
           (encode_int_vec left_bytes @ encode_int_vec right_bytes))
  | _ -> invalid_arg "Merkle node inputs must be canonical lowercase SHA3-256 hex digests"

let leaf commitment =
  digest (preimage merkle_leaf_tag (encode_int_vec commitment))

let empty width =
  if width < 0 then invalid_arg "empty width must be non-negative";
  digest (preimage merkle_empty_tag (encode_i64 width))

let node left right =
  match hex_to_bytes left, hex_to_bytes right with
  | Some left_bytes, Some right_bytes ->
      digest
        (preimage
           merkle_node_tag
           (encode_int_vec left_bytes @ encode_int_vec right_bytes))
  | _ -> invalid_arg "Merkle node inputs must be canonical lowercase SHA3-256 hex digests"

let all_same_width commitments width =
  List.for_all (fun c -> List.length c = width) commitments

let compress_level width level =
  let rec loop acc = function
    | [] -> List.rev acc
    | [left] -> List.rev (node left (empty width) :: acc)
    | left :: right :: rest -> loop (node left right :: acc) rest
  in
  match level with
  | [] -> []
  | [x] -> [x]
  | xs -> loop [] xs

let root commitments =
  let width =
    match commitments with
    | [] -> 0
    | first :: _ -> List.length first
  in
  if not (all_same_width commitments width) then
    invalid_arg "Merkle commitments must all have the same width";
  match commitments with
  | [] -> empty 0
  | _ ->
      let rec loop level =
        match level with
        | [] -> empty width
        | [x] -> x
        | xs -> loop (compress_level width xs)
      in
      loop (List.map leaf commitments)

let rec index_directions depth index =
  if depth <= 0 then []
  else (index mod 2 = 1) :: index_directions (depth - 1) (index / 2)

let path_root commitment siblings directions =
  let rec loop acc siblings directions =
    match siblings, directions with
    | [], [] -> Some acc
    | sibling :: rest_siblings, false :: rest_directions ->
        (try loop (node acc sibling) rest_siblings rest_directions with Invalid_argument _ -> None)
    | sibling :: rest_siblings, true :: rest_directions ->
        (try loop (node sibling acc) rest_siblings rest_directions with Invalid_argument _ -> None)
    | _ -> None
  in
  loop (leaf commitment) siblings directions

let membership_index ledger commitment =
  let rec loop index = function
    | [] -> None
    | x :: xs -> if x = commitment then Some index else loop (index + 1) xs
  in
  loop 0 ledger

let membership_prove ledger commitment =
  match membership_index ledger commitment with
  | None -> None
  | Some index ->
      let width =
        match ledger with
        | [] -> List.length commitment
        | first :: _ -> List.length first
      in
      if not (all_same_width ledger width) then
        invalid_arg "Merkle commitments must all have the same width";
      let rec loop current level siblings directions =
        match level with
        | [] -> None
        | [rt] ->
            Some
              {
                merkle_index = index;
                merkle_root = rt;
                merkle_siblings = List.rev siblings;
                merkle_directions = List.rev directions;
              }
        | _ ->
            let is_right = current mod 2 = 1 in
            let sibling =
              if is_right then List.nth level (current - 1)
              else if current + 1 < List.length level then List.nth level (current + 1)
              else empty width
            in
            loop
              (current / 2)
              (compress_level width level)
              (sibling :: siblings)
              (is_right :: directions)
      in
      loop index (List.map leaf ledger) [] []

let membership_verify commitment proof =
  proof.merkle_directions = index_directions (List.length proof.merkle_siblings) proof.merkle_index &&
  match hex_to_bytes proof.merkle_root, path_root commitment proof.merkle_siblings proof.merkle_directions with
  | Some _, Some rt -> rt = proof.merkle_root
  | _ -> false

File: isabella.hs/src/Canon/ZK/Confidential_Merkle.hs
-- | Cryptographic Merkle helpers for confidential note commitments.
--
-- This module is the Haskell runtime companion to
-- Canon/ZK/Authenticated_Merkle. It remains separate from
-- Confidential_Transaction while that verifier still uses the algebraic ledger
-- scaffold.
module Canon.ZK.Confidential_Merkle
  ( MerkleMembershipProof(..)
  , merkleDst
  , merkleLeafTag
  , merkleNodeTag
  , merkleEmptyTag
  , encodeLeaf
  , encodeEmpty
  , encodeNode
  , leaf
  , empty
  , node
  , root
  , pathRoot
  , membershipProve
  , membershipVerify
  ) where

import Data.Bits (shiftL, shiftR, (.&.), (.|.))
import Data.Char (ord)
import Data.List (findIndex)
import Data.Word (Word64, Word8)
import qualified Canon.ZK.Internal.RepeatedFS as RepeatedFS

data MerkleMembershipProof = MerkleMembershipProof
  { merkle_index :: Int
  , merkle_root :: String
  , merkle_siblings :: [String]
  , merkle_directions :: [Bool]
  }
  deriving (Eq, Show)

merkleDst :: String
merkleDst = "ISABELLA-CT-MERKLE-v1"

merkleLeafTag :: Int
merkleLeafTag = 0

merkleNodeTag :: Int
merkleNodeTag = 1

merkleEmptyTag :: Int
merkleEmptyTag = 2

word64LeBytes :: Word64 -> [Word8]
word64LeBytes value =
  [ fromIntegral ((value `shiftR` (8 * i)) .&. 0xff)
  | i <- [0 .. 7]
  ]

int64LeBytes :: Int -> [Word8]
int64LeBytes value =
  word64LeBytes (fromIntegral value :: Word64)

encodeIntVec :: [Int] -> [Word8]
encodeIntVec values =
  int64LeBytes (length values) ++ concatMap int64LeBytes values

preimage :: Int -> [Word8] -> [Word8]
preimage tag body =
  map (fromIntegral . ord) merkleDst ++ int64LeBytes tag ++ body

hexByte :: Word8 -> String
hexByte byte =
  let alphabet = "0123456789abcdef"
      hi = fromIntegral ((byte `shiftR` 4) .&. 0x0f)
      lo = fromIntegral (byte .&. 0x0f)
   in [alphabet !! hi, alphabet !! lo]

digestHex :: [Word8] -> String
digestHex =
  concatMap hexByte

digest :: [Word8] -> String
digest =
  digestHex . RepeatedFS.sha3_256

hexValue :: Char -> Maybe Word8
hexValue c
  | c >= '0' && c <= '9' = Just (fromIntegral (ord c - ord '0'))
  | c >= 'a' && c <= 'f' = Just (fromIntegral (10 + ord c - ord 'a'))
  | otherwise = Nothing

hexToBytes :: String -> Maybe [Word8]
hexToBytes hex
  | length hex /= 64 = Nothing
  | otherwise = go hex
 where
  go [] = Just []
  go (hi:lo:rest) = do
    hiValue <- hexValue hi
    loValue <- hexValue lo
    tailBytes <- go rest
    pure (((hiValue `shiftL` 4) .|. loValue) : tailBytes)
  go _ = Nothing

digestBytesToInts :: [Word8] -> [Int]
digestBytesToInts =
  map fromIntegral

encodeLeaf :: [Int] -> String
encodeLeaf commitment =
  digestHex (preimage merkleLeafTag (encodeIntVec commitment))

encodeEmpty :: Int -> String
encodeEmpty width
  | width < 0 = error "empty width must be non-negative"
  | otherwise = digestHex (preimage merkleEmptyTag (int64LeBytes width))

encodeNode :: String -> String -> String
encodeNode left right =
  case (hexToBytes left, hexToBytes right) of
    (Just leftBytes, Just rightBytes) ->
      digestHex
        (preimage
          merkleNodeTag
          (encodeIntVec (digestBytesToInts leftBytes) ++
           encodeIntVec (digestBytesToInts rightBytes)))
    _ -> error "Merkle node inputs must be canonical lowercase SHA3-256 hex digests"

leaf :: [Int] -> String
leaf commitment =
  digest (preimage merkleLeafTag (encodeIntVec commitment))

empty :: Int -> String
empty width
  | width < 0 = error "empty width must be non-negative"
  | otherwise = digest (preimage merkleEmptyTag (int64LeBytes width))

nodeMaybe :: String -> String -> Maybe String
nodeMaybe left right =
  case (hexToBytes left, hexToBytes right) of
    (Just leftBytes, Just rightBytes) ->
      Just
        (digest
          (preimage
            merkleNodeTag
            (encodeIntVec (digestBytesToInts leftBytes) ++
             encodeIntVec (digestBytesToInts rightBytes))))
    _ -> Nothing

node :: String -> String -> String
node left right =
  case nodeMaybe left right of
    Just value -> value
    Nothing -> error "Merkle node inputs must be canonical lowercase SHA3-256 hex digests"

sameWidth :: [[Int]] -> Int -> Bool
sameWidth commitments width =
  all ((== width) . length) commitments

compressLevel :: Int -> [String] -> [String]
compressLevel _ [] = []
compressLevel _ [x] = [x]
compressLevel width xs =
  go xs
 where
  go [] = []
  go [left] = [node left (empty width)]
  go (left:right:rest) = node left right : go rest

root :: [[Int]] -> String
root commitments =
  let width = case commitments of
        [] -> 0
        first:_ -> length first
   in if not (sameWidth commitments width)
        then error "Merkle commitments must all have the same width"
        else case commitments of
          [] -> empty 0
          _ -> go (map leaf commitments)
            where
              go [] = empty width
              go [x] = x
              go xs = go (compressLevel width xs)

indexDirections :: Int -> Int -> [Bool]
indexDirections depth index
  | depth <= 0 = []
  | otherwise = odd index : indexDirections (depth - 1) (index `div` 2)

pathRoot :: [Int] -> [String] -> [Bool] -> Maybe String
pathRoot commitment siblings directions =
  go (leaf commitment) siblings directions
 where
  go acc [] [] = Just acc
  go acc (sibling:restSiblings) (False:restDirections) = do
    parent <- nodeMaybe acc sibling
    go parent restSiblings restDirections
  go acc (sibling:restSiblings) (True:restDirections) = do
    parent <- nodeMaybe sibling acc
    go parent restSiblings restDirections
  go _ _ _ = Nothing

membershipProve :: [[Int]] -> [Int] -> Maybe MerkleMembershipProof
membershipProve ledger commitment = do
  index <- findIndex (== commitment) ledger
  let width = case ledger of
        [] -> length commitment
        first:_ -> length first
  if not (sameWidth ledger width)
    then Nothing
    else go index width index (map leaf ledger) [] []
 where
  go _ _ _ [] _ _ = Nothing
  go originalIndex _ _ [rt] siblings directions =
    Just
      MerkleMembershipProof
        { merkle_index = originalIndex
        , merkle_root = rt
        , merkle_siblings = reverse siblings
        , merkle_directions = reverse directions
        }
  go originalIndex width current level siblings directions =
    let isRight = odd current
        sibling =
          if isRight
            then level !! (current - 1)
            else if current + 1 < length level
              then level !! (current + 1)
              else empty width
     in go
          originalIndex
          width
          (current `div` 2)
          (compressLevel width level)
          (sibling : siblings)
          (isRight : directions)

membershipVerify :: [Int] -> MerkleMembershipProof -> Bool
membershipVerify commitment proof =
  merkle_directions proof == indexDirections (length (merkle_siblings proof)) (merkle_index proof) &&
  case (hexToBytes (merkle_root proof), pathRoot commitment (merkle_siblings proof) (merkle_directions proof)) of
    (Just _, Just rt) -> rt == merkle_root proof
    _ -> False
